src/shared/qmlpath.cpp

```diff
@@ -0,0 +1,287 @@
+        number.append(*input);
+        ++input;
+        --size;
+      }
+
+      if (size == 0 || *input != ']') return false;
+
+      ++input;
+      --size;
+
+      Filter filter;
+      filter.m_type = Filter::Index;
+
+      bool ok = false;
+      filter.m_index = number.toInt(&ok);
+      if (!ok || filter.m_index < 0) return false;
+
+      data.m_filters.append(filter);
+      continue;
+    }
+
+    if (*input == '{') {
+      ++input;
+      --size;
+
+      QString propertyValue;
+      while (size && *input != '}') {
+        propertyValue.append(*input);
+        ++input;
+        --size;
+      }
+
+      if (size == 0 || *input != '}') return false;
+
+      ++input;
+      --size;
+
+      QStringList parts = propertyValue.split('=');
+      if (parts.length() > 2) return false;
+
+      Filter filter;
+      filter.m_type = Filter::Property;
+
+      filter.m_property = parts[0].toLocal8Bit();
+      if (filter.m_property.isEmpty()) return false;
+
+      if (parts.length() > 1) {
+        filter.m_hasPropertyValue = true;
+        filter.m_propertyValue = parts[1];
+      }
+
+      data.m_filters.append(filter);
+      continue;
+    }
+
+    return false;
+  }
+
+  blocks.append(data);
+  return true;
+}
+
+QQuickItem* QmlPath::evaluate(QQmlApplicationEngine* engine) const {
+  if (!engine) {
+    return nullptr;
+  }
+
+  if (m_blocks.isEmpty()) {
+    return nullptr;
+  }
+
+  QList<QQuickItem*> list;
+  for (QObject* object : engine->rootObjects()) {
+    QQuickItem* item = qobject_cast<QQuickItem*>(object);
+    if (item) {
+      list.append(item);
+    } else {
+      for (QObject* child : object->children()) {
+        QQuickItem* item = qobject_cast<QQuickItem*>(child);
+        if (item) {
+          list.append(item);
+        }
+      }
+    }
+  }
+
+  return evaluateItems(nullptr, list, m_blocks.cbegin());
+}
+
+QQuickItem* QmlPath::evaluateItems(QQuickItem* currentItem,
+                                   const QList<QQuickItem*>& items,
+                                   QList<Data>::const_iterator i) const {
+  if (i == m_blocks.cend()) {
+    return currentItem;
+  }
+
+  if (items.isEmpty()) {
+    return nullptr;
+  }
+
+  QList<QQuickItem*> results;
+
+  if (i->m_key.isEmpty()) {
+    results.append(items);
+  } else {
+    for (QQuickItem* item : items) {
+      if (item->objectName() == i->m_key) {
+        results.append(item);
+      }
+    }
+
+    if (i->m_nested) {
+      for (QQuickItem* item : items) {
+        results.append(findItems(item, i->m_key));
+      }
+    }
+  }
+
+  for (const Filter& filter : i->m_filters) {
+    switch (filter.m_type) {
+      case Filter::Index:
+        results = filterByIndex(results, filter);
+        break;
+
+      case Filter::Property:
+        filterByProperty(results, filter);
+        break;
+    }
+  }
+
+  if (results.isEmpty()) {
+    return nullptr;
+  }
+
+  return evaluateItems(results[0], collectChildItems(results[0]), i + 1);
```

Comment:
Perhaps I am overthinking this, but am getting the impression that this can prevent complicated matches from working as expected. For example, if we consider an example similar to the syntax blurb in the header file: `/abc//foo{p=42}/bar`

And give it the following QML:
```
Item {
  objectName: "abc"
  Item {
    objectName: "foo"
    property p: 42
    Item {
      objectName: "not-bar"
    }
  }
  Item {
    objectName: "example"
    Item {
      objectName: "foo"
      property p: 42
      Item {
       objectName: "bar"
      }
   }
}
```

Then presumeably, we will wind up with both `foo` objects in `results` but by only evaluating the first entry in the list we never actually find the one which satisfies the full path being filtered against. I think this needs to iterate on `results` until either a match is found or the list is exhausted.

Reply:
And when considering that nesting can result in multiple matches, it's probably worth documenting somewhere exactly which item will be found (IIUC, I think we get the first match found in a pre-order traversal of the QML tree).

Reply:
The first implementation was like you described. But this was incompatible with the 'range' implementation.
Let's use your QML tree. If I do: `/abc//foo[1]` I pick the nested foo. `/abc/foo[0]`, picks the first one. If you don't specify the index, the first one is evaluated. Basically, the current logic is that we have an implicit `[0]` at the end of each path.

Reply:
Actually, you are right. Fixed.